ludwig/features/image_feature.py

```diff
@@ -166,110 +178,123 @@ def add_feature_data(
+            'num_processes',
+            preprocessing_parameters['num_processes']
+        )
+        csv_path = None
+        if hasattr(dataset_df, 'csv'):
+            csv_path = os.path.dirname(os.path.abspath(dataset_df.csv))
+
+        num_images = len(dataset_df)
+        if num_images == 0:
+            raise ValueError('There are no images in the dataset provided.')
+
+        first_image_path = dataset_df[feature['name']][0]
+        if csv_path is None and not os.path.isabs(first_image_path):
+            raise ValueError('Image file paths must be absolute')
+
+        first_image_path = get_abs_path(csv_path, first_image_path)
+
+        (should_resize,
+         width,
+         height,
+         num_channels,
+         user_specified_num_channels,
+         first_image) = \
+            ImageBaseFeature._finalize_preprocessing_parameters(
+                preprocessing_parameters, first_image_path
+            )
 
         metadata[feature['name']]['preprocessing']['height'] = height
         metadata[feature['name']]['preprocessing']['width'] = width
         metadata[feature['name']]['preprocessing'][
             'num_channels'] = num_channels
 
+        read_image_and_resize = partial(
+            ImageBaseFeature._read_image_and_resize,
+            img_width=width,
+            img_height=height,
+            should_resize=should_resize,
+            num_channels=num_channels,
+            resize_method=preprocessing_parameters['resize_method'],
+            user_specified_num_channels=user_specified_num_channels
+        )
+        all_file_paths = [get_abs_path(csv_path, file_path)
+                          for file_path in dataset_df[feature['name']]]
+
         if feature['preprocessing']['in_memory']:
+            # Number of processes to run in parallel for preprocessing
+            num_processes = feature['preprocessing']['num_processes']
+            metadata[feature['name']]['preprocessing'][
+                'num_processes'] = num_processes
+
             data[feature['name']] = np.empty(
                 (num_images, height, width, num_channels),
                 dtype=np.uint8
             )
-            for i in range(len(dataset_df)):
-                filepath = get_abs_path(
-                    csv_path,
-                    dataset_df[feature['name']][i]
-                )
-
-                img = ImageBaseFeature._read_image_and_resize(
-                    filepath,
-                    width,
-                    height,
-                    should_resize,
-                    num_channels,
-                    preprocessing_parameters['resize_method'],
-                    user_specified_num_channels
+            with Pool(num_processes) as pool:
+                logger.warning('Using {} processes for preprocessing '
+                               'images'.format(num_processes))
+                data[feature['name']] = np.array(
+                    pool.map(read_image_and_resize, all_file_paths)
```

Comment:
Nit: closing bracket on a new line

Reply:
It looks correct to me, there are 2 open brackets at the end of the line, one of map and one of array, the array one is at the next line